templates/_worker.js/matcher.ts

```diff
@@ -0,0 +1,414 @@
+import { parse } from 'cookie';
+import type { MatchPCREResult, MatchedSet } from './utils';
+import {
+	applyHeaders,
+	applyPCREMatches,
+	applySearchParams,
+	getNextPhase,
+	isUrl,
+	runOrFetchBuildOutputItem,
+	checkRouteMatch as checkRouteMatchUtility,
+} from './utils';
+
+export type CheckRouteStatus = 'skip' | 'next' | 'done' | 'error';
+export type CheckPhaseStatus = Extract<CheckRouteStatus, 'error' | 'done'>;
+
+export class Matcher {
+	/** Processed routes from the Vercel build output config. */
+	private routes: ProcessedVercelRoutes;
+	/** Vercel build output. */
+	private output: VercelBuildOutput;
+	/** Static assets fetcher. */
+	private assets: Fetcher;
+	/** Execution context. */
+	private ctx: ExecutionContext;
+
+	/** Request to match */
+	private req: Request;
+	/** URL from the request to match */
+	private url: URL;
+	/** Cookies from the request to match */
+	private cookies: Record<string, string>;
+
+	/** Path for the matched route */
+	public path: string;
+	/** Status for the response object */
+	public status: number | undefined;
+	/** Headers for the response object */
+	public headers: { normal: Headers; important: Headers };
+	/** Search params for the response object */
+	public searchParams: URLSearchParams;
+
+	/**
+	 * Creates a new instance of a request matcher.
+	 *
+	 * The matcher is used to match a request to a route and run the route's middleware.
+	 *
+	 * @param config The processed Vercel build output config.
+	 * @param output Vercel build output.
+	 * @param assets Static assets fetcher.
+	 * @param ctx Execution context.
+	 * @param req Request object.
+	 * @param prevMatch The previous match from a routing phase to initialize the matcher with.
+	 * @returns The matched set of path, status, headers, and search params.
+	 */
+	constructor(
+		routes: ProcessedVercelRoutes,
+		output: VercelBuildOutput,
+		assets: Fetcher,
+		ctx: ExecutionContext,
+		req: Request,
+		prevMatch?: MatchedSet
+	) {
+		this.routes = routes;
+		this.output = output;
+		this.assets = assets;
+		this.ctx = ctx;
+
+		this.req = req;
+		this.url = new URL(req.url);
+		this.cookies = parse(req.headers.get('cookie') || '');
+
+		this.path = prevMatch?.path || this.url.pathname || '/';
+		this.status = prevMatch?.status;
+		this.headers = prevMatch?.headers || {
+			normal: new Headers(),
+			important: new Headers(),
+		};
+		this.searchParams = prevMatch?.searchParams || new URLSearchParams();
+	}
+
+	/**
+	 * Checks if a Vercel source route from the build output config matches the request.
+	 *
+	 * @param route Build output config source route.
+	 * @param checkStatus Whether to check the status code of the route.
+	 * @returns The source path match result if the route matches, otherwise `undefined`.
+	 */
+	private checkRouteMatch(
+		route: VercelSource,
+		checkStatus?: boolean
+	): MatchPCREResult | undefined {
+		return checkRouteMatchUtility(route, this.path, {
```

Comment:
If I move the function inside the class, then I'll delete the tests that are written specifically for the function because, well, I'm not making it a static method, and it's basically tested as part of the router tests anyway. If you're fine with that, I can.

Reply:
I ended up adapting the tests to use the new router test suite instead. 👍🏻